google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -154,64 +136,85 @@ private void forget() {
+      pendingNacks.add(this.ackRequestData);
       forget();
     }
 
     @Override
     public void onSuccess(AckReply reply) {
-      LinkedBlockingQueue<String> destination;
       switch (reply) {
         case ACK:
-          destination = pendingAcks;
+          pendingAcks.add(this.ackRequestData);
           // Record the latency rounded to the next closest integer.
           ackLatencyDistribution.record(
               Ints.saturatedCast(
                   (long) Math.ceil((clock.millisTime() - receivedTimeMillis) / 1000D)));
           break;
         case NACK:
-          destination = pendingNacks;
+          pendingNacks.add(this.ackRequestData);
           break;
         default:
           throw new IllegalArgumentException(String.format("AckReply: %s not supported", reply));
       }
-      destination.add(ackId);
       forget();
     }
   }
 
   interface AckProcessor {
-    void sendAckOperations(
-        List<String> acksToSend, List<PendingModifyAckDeadline> ackDeadlineExtensions);
+    public void sendAckOperations(List<AckRequestData> ackRequestDataList);
+
+    public void sendModackOperations(List<ModackRequestData> modackRequestDataList);
   }
 
-  MessageDispatcher(
-      MessageReceiver receiver,
-      AckProcessor ackProcessor,
-      Duration ackExpirationPadding,
-      Duration maxAckExtensionPeriod,
-      Duration maxDurationPerAckExtension,
-      Distribution ackLatencyDistribution,
-      FlowController flowController,
-      Executor executor,
-      ScheduledExecutorService systemExecutor,
-      ApiClock clock) {
-    this.executor = executor;
-    this.systemExecutor = systemExecutor;
-    this.ackExpirationPadding = ackExpirationPadding;
-    this.maxAckExtensionPeriod = maxAckExtensionPeriod;
-    this.maxSecondsPerAckExtension = Math.toIntExact(maxDurationPerAckExtension.getSeconds());
-    this.receiver = receiver;
-    this.ackProcessor = ackProcessor;
-    this.flowController = flowController;
-    // 601 buckets of 1s resolution from 0s to MAX_ACK_DEADLINE_SECONDS
-    this.ackLatencyDistribution = ackLatencyDistribution;
+  private MessageDispatcher(Builder builder) {
+    executor = builder.executor;
+    systemExecutor = builder.systemExecutor;
+    ackExpirationPadding = builder.ackExpirationPadding;
+    maxAckExtensionPeriod = builder.maxAckExtensionPeriod;
+
+    minDurationPerAckExtensionSeconds =
+        Math.toIntExact(builder.minDurationPerAckExtension.getSeconds());
+    minDurationPerAckExtensionDefaultUsed = builder.minDurationPerAckExtensionDefaultUsed;
+    maxDurationPerAckExtensionSeconds =
+        Math.toIntExact(builder.maxDurationPerAckExtension.getSeconds());
+    maxDurationPerAckExtensionDefaultUsed = builder.maxDurationPerAckExtensionDefaultUsed;
+
+    // Start the deadline at the minimum ack deadline so messages which arrive before this is
+    // updated will not have a long ack deadline.
+    messageDeadlineSeconds.set(minDurationPerAckExtensionSeconds);
```

Comment:
I think this will change current behavior. The default minDurationPerAckExtensionSeconds is 0. I think we only want to set this to the minDurationPerAckExtension if the default is not used. If the default is used, we should set this to the MIN_ACK_DEADLINE_SECONDS (or whatever this is named now)

Reply:
If this is not set in the `Subscriber`, the default will be used and it will propagate down to the `MessageDispatcher`